### DISCUSSION TITLE
Questions about lines and LineRender pass

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#564] [General]
I am trying to get more control over the Lines, and understand what is actually going on in the LineRender node of the Render Layer. My current understanding is:

1. Lines are Detected in the Main Pass using the Normal Depth and ID textures from the PrePass
2. Lines are Colored and Expanded using Line Color and Line Width from the MainPass
3. Lines are composited over the Color from the main pass.

Is that correct?

What I would like is to be able to make changes between some of these steps, such as changing the detected lines before they are expanded, and keeping the MainPass Color and Lines as separate layers that aren't composited within the Render Layer so that Lines can be saved as their own pass, and so that other post processes can take place on the main image before the lines are composited.

I have several questions as I try to understand my options:

1) I have found that I can insert a Screen pass between the MainPass and LineRender to make changes to the Color and Width. I'm not sure I would ever need to do this specifically since most things it can do are probably also doable in the Main Pass in the first place. But if there was something that needed to be done in Screen Space instead of the MainPass, this seems to work. But I assume it is very inefficient compared to doing it within the LineRender itself, as I assume that is also a screen pass of sorts?
![image](https://github.com/bnpr/Malt/assets/36803956/6ae9fb81-a201-42ac-ba97-5913043e3df1)

2) I have tried reading the code for the LineRender (python), and various of the line functions like LineDetection (glsl). I generally can grasp what is going on inside these functions, but I haven't been able to find what is calling them, so I haven't been able to grasp the bigger picture. This is undoubtably because I am a noob at understanding code structures. Can you point me towards the correct file(s)?

3) Would it be possible to rebuild the LineRender out of existing nodes within a Screen Pass? I am assuming that LineRender is mostly the internal Line Expand node. I can make a screen pass and put that node inside like this, but I don't know how to use its Result output socket, and I'm not sure if the inputs are explicitly filling. I assume the 

![image](https://github.com/bnpr/Malt/assets/36803956/7b35fec0-b329-4d41-8eec-c885bc71f477)
![image](https://github.com/bnpr/Malt/assets/36803956/a3f16a10-881b-4c31-90d1-21b14093de6b)

4) Related to 3, all these passes are inputting Normal Depth and ID textures (they seem to implicitly input even if not hooked up.) But how do I actually access these textures within the passes? I know they are being used by various nodes, but is there a way to input the textures themselves with a node? The other input textures show up on an Input node, but not these. And like the ID can be accessed within the MainPass with the ID input node. But this node doesn't work within ScreenPass.

5) What exactly is contained within Scene?

6) I remember in older versions of Malt before the newer nodes, it was possible to get info from the PrePass as a texture in the MainPass. This let you do things like adjust the screen UV coords of some input date the same way you can in a Screen Pass. I'm guessing that is all hidden now, but is it still possible somewhere without digging into the code?

Thanks as always, and sorry if these are silly questions, but I'm not advanced enough at reading the code to answer these things for myself :(

**Comment by pragma37** [comment]
> My current understanding is: (...) Is that correct?

Yes!

> keeping the MainPass Color and Lines as separate layers that aren't composited within the Render Layer so that Lines can be saved as their own pass, and so that other post processes can take place on the main image before the lines are composited.

You can input whatever you want into Line Render color, it doesn't have to come directly from Main Pass color. You can even input a zero alpha texture to get only the expanded lines on its own texture.

1. Malt is not super efficient performance-wise in the first place, so if you need to do something like that I'd say go for it, the system is designed to allow it.

2. The LineRender node is executed from the generated Render Layers script. Inside your `.malt-autogenerated` folder there should be something like `<filename>-<render layer nodetree name>-Render Layer.py`.
But I don't think you need that to understand what's going on with `Line Render.py`, it just receives some textures, do its thing, and output other textures. The general flow is exactly what it looks like in the nodes themselves.

3. Yes, that should be roughly the same (see `Passes/LineComposite.glsl`, that's what the LineRender node uses).
You're only missing compositing the 2 layers together, and disabling the `Layer Only` option so lines can go out of the mesh shape.

4. The textures themselves are not exposed directly, but the input position, normal and ID and extracted from it.
You should also be able to put `IN_NORMAL_DEPTH` in an inline code node and connect that to a sampler2d socket.
Note at IDs are usampler2d, so `IN_ID` can't be used with nodes that take a sampler2d socket.

>  And like the ID can be accessed within the MainPass with the ID input node. But this node doesn't work within ScreenPass.

This should work.

5. Literally the whole scene data (including loaded meshes and materials with their shaders and textures), and the shader callbacks that other nodes add.
See `Malt/Scene.py`.

6. You don't need to dig into the code, just add a custom output socket to the pre-pass node and a custom input socket to the main-pass node.

> Thanks as always, and sorry if these are silly questions

You're welcome! They're not silly questions. 👍

**Reply by aVersionOfReality** [reply]
Excellent, I got it working. Thank you!

**Comment by aVersionOfReality** [comment]
Some other Line Questions:

1. Is there a way to get an SDF or gradient from the center to the edge of a line? I want to use some thick lines to get masks from the edges of things into their interiors. Blurring them might work too, but Lines already have more controls about where they are placed.
2. How are World line width units calculated? I want to make other effects that are also consistent in size relative to World Space. For example a Blue/Glow effect on a LED on a character's outfit should always be the same size relative to the character. I think I know how this is done, but since such a system already exists for lines, I don't want to reinvent the wheel.

**Reply by pragma37** [reply]
1. There's no built-in way. The "right" way to do it would be to make a copy of the `line_expand` function and store the `offset_length / offset_width` in the `if (override)` case.  You can add the result as an output parameter.
2. See `line_width_2`:
```glsl
if(width_units == 2)//World
{
    width_scale /= pixel_world_size() * 100.0;
}
```
Pixel world size is also available as a node.

**Reply by aVersionOfReality** [reply]
Oh great, I'll check those out.
